packages/@sanity/form-builder/src/sanity/inputs/reference/SanityArrayItemReferenceInput.tsx

```diff
@@ -0,0 +1,200 @@
+import {FormFieldPresence} from '@sanity/base/presence'
+import {from, throwError} from 'rxjs'
+import {catchError, mergeMap} from 'rxjs/operators'
+import withValuePath from '../../../utils/withValuePath'
+import withDocument from '../../../utils/withDocument'
+import {useReferenceInputOptions} from '../../contexts/ReferenceInputOptions'
+import PatchEvent from '../../../PatchEvent'
+import * as adapter from '../client-adapters/reference'
+import {ArrayItemReferenceInput} from '../../../inputs/ReferenceInput/ArrayItemReferenceInput'
+import {EditReferenceEvent} from '../../../inputs/ReferenceInput/types'
+import {schema} from '../../../legacyParts'
+
+// eslint-disable-next-line require-await
+async function resolveUserDefinedFilter(
+  options: ReferenceOptions | undefined,
+  document: SanityDocument,
+  valuePath: Path
+): Promise<ReferenceFilterSearchOptions> {
+  if (!options) {
+    return {}
+  }
+
+  if (typeof options.filter === 'function') {
+    const parentPath = valuePath.slice(0, -1)
+    const parent = get(document, parentPath) as Record<string, unknown>
+    return options.filter({document, parentPath, parent})
+  }
+
+  return {
+    filter: options.filter,
+    params: 'filterParams' in options ? options.filterParams : undefined,
+  }
+}
+
+export type Props = {
+  value: Reference
+  compareValue?: Reference
+  type: ReferenceSchemaType
+  markers: Marker[]
+  focusPath: Path
+  readOnly?: boolean
+  isSortable: boolean
+  onFocus: (path: Path) => void
+  onChange: (event: PatchEvent) => void
+  level: number
+  presence: FormFieldPresence[]
+
+  // From withDocument
+  document: SanityDocument
+
+  // From withValuePath
+  getValuePath: () => Path
+}
+
+function useValueRef<T>(value: T): {current: T} {
+  const ref = useRef(value)
+  ref.current = value
+  return ref
+}
+
+type SearchError = {
+  message: string
+  details?: {
+    type: string
+    description: string
+  }
+}
+
+const SanityReferenceInput = forwardRef(function SanityReferenceInput(
+  props: Props,
+  ref: ForwardedRef<HTMLInputElement>
+) {
+  const {getValuePath, type, document} = props
+  const {
+    EditReferenceLinkComponent,
+    onEditReference,
+    activePath,
+    initialValueTemplateItems,
+  } = useReferenceInputOptions()
+
+  const documentRef = useValueRef(document)
+
+  const valuePath = useMemo(getValuePath, [getValuePath])
+
+  const documentTypeName = documentRef.current?._type
+
+  const isDocumentLiveEdit = useMemo(() => {
+    return schema.get(documentTypeName).liveEdit
+  }, [documentTypeName])
+
+  const disableNew = type.options?.disableNew === true
+
+  const handleSearch = useCallback(
+    (searchString: string) =>
+      from(resolveUserDefinedFilter(type.options, documentRef.current, getValuePath())).pipe(
+        mergeMap(({filter, params}) =>
+          adapter.search(searchString, type, {
+            ...type.options,
+            filter,
+            params,
+            tag: 'search.reference',
+          })
```

Comment:
are you missing a limit here?

Reply:
added a hard coded limit of `50` hits to the search function. I think we can add support for customising the limit in a separate pass.